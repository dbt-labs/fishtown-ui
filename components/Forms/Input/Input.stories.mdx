import { Meta, ArgsTable, Story, Preview, Canvas } from '@storybook/addon-docs/blocks';

import { Input } from './Input';
import { Field } from '../Field';

<Meta title="Components/Forms/Input" component={Input} />

# Input

A basic form element, accepts any standard type, passes through any additional valid html props.

## Props (any valid HTML Prop)

<ArgsTable of={Input} />

## Examples

### Basic `Input`

<Canvas>
  <Story
    name="Basic"
    args={{
      id: undefined,
    }}
  >
    {(args) => <Input {...args} />}
  </Story>
</Canvas>

### Usage with `Field`

<Canvas>
  <Story name="Field">
    <Field fieldId="text-field" label="Text Field" isEdit>
      {(innerProps) => <Input {...innerProps} placeholder="I'm a text field!" />}
    </Field>
  </Story>
</Canvas>

### Placeholder Input

<Canvas>
  <Story
    name="Placeholder"
    args={{
      id: undefined,
      placeholder: 'Enter some text...',
    }}
  >
    {(args) => <Input {...args} />}
  </Story>
</Canvas>

### Disabled Input

<Canvas>
  <Story
    name="Disabled"
    args={{
      id: undefined,
      placeholder: 'This input is disabled',
      disabled: true,
    }}
  >
    {(args) => <Input {...args} />}
  </Story>
</Canvas>

### Input with value

<Canvas>
  <Story
    name="Value"
    args={{
      id: undefined,
      placeholder: 'Enter some text',
      defaultValue: 'I have a value!',
    }}
  >
    {(args) => <Input {...args} />}
  </Story>
</Canvas>

### Extra Input types

<Canvas>
  <Story
    name="Types"
    args={{
      types: ['password', 'number', 'email', 'hidden', 'search', 'tel'],
    }}
  >
    {(args) =>
      args.types.map((type) => (
        <div key={type} className="tw-mt-2 tw-mb-2">
          <Input
            id={`${type}-input`}
            type={type}
            key={type}
            placeholder={`${type} Input`}
          />
        </div>
      ))
    }
  </Story>
</Canvas>
