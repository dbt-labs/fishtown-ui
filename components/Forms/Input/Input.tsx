import React, { forwardRef, ForwardRefRenderFunction } from 'react';
import cx from 'classnames';
import { v4 as uuidv4 } from 'uuid';

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  type?: 'text' | 'password' | 'number' | 'email' | 'hidden' | 'search' | 'tel';
}

const ForwardInput: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  { type = 'text', id = `field_${uuidv4()}`, className, ...restProps }: InputProps,
  ref: React.MutableRefObject<HTMLInputElement>
): React.ReactElement => {
  const classNames = cx('fui-field-row__input', className);
  return (
    <>
      <input ref={ref} className={classNames} id={id} type={type} {...restProps} />
    </>
  );
};

export const Input = forwardRef(ForwardInput);
