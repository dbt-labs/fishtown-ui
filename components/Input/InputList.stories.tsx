import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { InputList, InputListProps } from './InputList';

export default {
  title: 'Components/InputList',
  component: InputList,
} as Meta;

const Template: Story<InputListProps> = (args) => <InputList {...args} />;
Template.argTypes = {};

export const Basic = Template.bind({});
Basic.args = {
    values: ["value 1", "value 2", "value 3"],
    isDisabled: false,
};