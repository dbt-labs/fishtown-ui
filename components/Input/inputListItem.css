@layer components {
    .fui-input-list-item {
        @apply tw-pb-2;
    }

    .fui-input-list-item__field {
        @apply tw-pl-2 tw-bg-gray-lighter tw-rounded tw-h-6 tw-w-56
    }

    .fui-input-list-item__delete {
        @apply tw-p-2 tw-fill-current tw-text-gray-dark
    }

    .fui-input-list-item__delete:hover {
        @apply tw-fill-current tw-text-red
    }
}