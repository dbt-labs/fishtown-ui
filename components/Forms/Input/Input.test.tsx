import React from 'react';
import { render } from '@testing-library/react';
import fireEvent from '@testing-library/user-event';

import { Input } from './Input';

describe('Component Input', () => {
  test('should render an input defaulted to text', () => {
    const { container } = render(<Input id="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should pass through any supported html input props', () => {
    const ref = React.createRef<HTMLInputElement>();
    const { container } = render(
      <Input
        name="test"
        id="test"
        type="text"
        disabled={true}
        aria-label="test"
        data-test="test"
        ref={ref}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(ref.current?.getAttribute('id')).toBe('test');
  });

  test('should call onChange callbacks', () => {
    const handleChange = jest.fn();
    const { getByTestId } = render(
      <Input id="test" data-testid="test" onChange={handleChange} />
    );
    fireEvent.type(getByTestId('test'), 'test');
    expect(handleChange).toHaveBeenCalled();
  });
});
