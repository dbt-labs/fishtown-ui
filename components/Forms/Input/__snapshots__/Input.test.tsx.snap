// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Input should pass through any supported html input props 1`] = `
<input
  aria-label="test"
  class="fui-field-row__input"
  data-test="test"
  disabled=""
  id="test"
  name="test"
  type="text"
/>
`;

exports[`Component Input should render an input defaulted to text 1`] = `
<input
  class="fui-field-row__input"
  id="test"
  type="text"
/>
`;
